False positive for `attributes` rule when using `@MainActor`

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

It seems using `@MainActor` triggers a false positive depending on where `@MainActor` is exactly. See example below.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
Linting Swift files in current working directory
Linting 'test.swift' (1/1)
/Users/marco/Desktop/test/test.swift:5:7: warning: Attributes Violation: Attributes should be on their own lines in functions and types, but on the same line as variables and imports. (attributes)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version 0.50.1
* Installation method used: Homebrew
* Paste your configuration file:

```yml
opt_in_rules: 
  - attributes
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  > No
* Which Xcode version are you using (check `xcodebuild -version`)?
  > Xcode 14.1
  > Build version 14B47b

```swift
// This triggers a violation:
import AppKit

@NSApplicationMain
@MainActor
final class AppDelegate: NSAppDelegate {}
```

The order of the attributes is important:
```swift
// This does NOT trigger a violation:
import AppKit

@MainActor
@NSApplicationMain
final class AppDelegate: NSAppDelegate {}
```

I also tried using `@main` instead of `@NSApplicationMain` with the same result.

Interestingly, using e.g. `@main` _and_ `@NSApplicationMain` but _without_ `@MainActor` does not trigger a warning, regardless of the order. So it seems like the problem is `@MainActor`.

--- a/.swiftlint.yml
+++ b/.swiftlint.yml
@@ -78,6 +78,9 @@ opt_in_rules:
   - xct_specific_matcher
   - yoda_condition
 
+attributes:
+  always_on_line_above:
+    - "@OptionGroup"
 identifier_name:
   excluded:
     - id

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -49,6 +49,10 @@
   [SimplyDanny](https://github.com/SimplyDanny)
   [#4645](https://github.com/realm/SwiftLint/issues/4645)
 
+* Consider custom attributes in `attributes` rule.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#4599](https://github.com/realm/SwiftLint/issues/4599)
+
 * Fix false positives on `private_subject` rule when using
   subjects inside functions.  
   [Marcelo Fabri](https://github.com/marcelofabri)

--- a/Source/SwiftLintFramework/Rules/Style/AttributesRule.swift
+++ b/Source/SwiftLintFramework/Rules/Style/AttributesRule.swift
@@ -121,7 +121,7 @@ private struct RuleHelper {
 
     func hasViolation(
         locationConverter: SourceLocationConverter,
-        attributesAndPlacements: [(AttributeSyntax, AttributePlacement)]
+        attributesAndPlacements: [(SyntaxProtocol, AttributePlacement)]
     ) -> Bool {
         var linesWithAttributes: Set<Int> = [keywordLine]
         for (attribute, placement) in attributesAndPlacements {
@@ -147,23 +147,69 @@ private struct RuleHelper {
     }
 }
 
+private enum Attribute {
+    case builtIn(AttributeSyntax)
+    case custom(CustomAttributeSyntax)
+
+    static func from(syntax: SyntaxProtocol) -> Self? {
+        if let attribute = syntax.as(AttributeSyntax.self) {
+            return .builtIn(attribute)
+        }
+        if let attribute = syntax.as(CustomAttributeSyntax.self) {
+            return .custom(attribute)
+        }
+        return nil
+    }
+
+    var hasArguments: Bool {
+        switch self {
+        case let .builtIn(attribute):
+            return attribute.argument != nil
+        case let .custom(attribute):
+            return attribute.argumentList != nil
+        }
+    }
+
+    var name: String? {
+        switch self {
+        case let .builtIn(attribute):
+            return attribute.attributeName.text
+        case let .custom(attribute):
+            return attribute.attributeName.as(SimpleTypeIdentifierSyntax.self)?.typeName
+        }
+    }
+
+    var syntaxNode: SyntaxProtocol {
+        switch self {
+        case let .builtIn(attribute):
+            return attribute
+        case let .custom(attribute):
+            return attribute
+        }
+    }
+}
+
 private extension AttributeListSyntax {
     func attributesAndPlacements(configuration: AttributesConfiguration, shouldBeOnSameLine: Bool)
-        -> [(AttributeSyntax, AttributePlacement)] {
-        self
-            .children(viewMode: .sourceAccurate)
-            .compactMap { $0.as(AttributeSyntax.self) }
-            .map { attribute in
-                let atPrefixedName = "@\(attribute.attributeName.text)"
+        -> [(SyntaxProtocol, AttributePlacement)] {
+        self.children(viewMode: .sourceAccurate)
+            .compactMap { Attribute.from(syntax: $0) }
+            .compactMap { attribute in
+                guard let attributeName = attribute.name else {
+                    return nil
+                }
+                let atPrefixedName = "@\(attributeName)"
                 if configuration.alwaysOnSameLine.contains(atPrefixedName) {
-                    return (attribute, .sameLineAsDeclaration)
+                    return (attribute.syntaxNode, .sameLineAsDeclaration)
                 } else if configuration.alwaysOnNewLine.contains(atPrefixedName) {
-                    return (attribute, .dedicatedLine)
-                } else if attribute.argument != nil {
-                    return (attribute, .dedicatedLine)
+                    return (attribute.syntaxNode, .dedicatedLine)
+                } else if attribute.hasArguments {
+                    return (attribute.syntaxNode, .dedicatedLine)
                 }
 
-                return shouldBeOnSameLine ? (attribute, .sameLineAsDeclaration) : (attribute, .dedicatedLine)
+                return shouldBeOnSameLine
+                    ? (attribute.syntaxNode, .sameLineAsDeclaration)
+                    : (attribute.syntaxNode, .dedicatedLine)
             }
     }
 

--- a/Source/SwiftLintFramework/Rules/Style/AttributesRuleExamples.swift
+++ b/Source/SwiftLintFramework/Rules/Style/AttributesRuleExamples.swift
@@ -73,6 +73,13 @@ internal struct AttributesRuleExamples {
         func refreshable(action: @escaping @Sendable () async -> Void) -> some View {
             modifier(RefreshableModifier(action: action))
         }
+        """),
+        Example("""
+        import AppKit
+
+        @NSApplicationMain
+        @MainActor
+        final class AppDelegate: NSAppDelegate {}
         """)
     ]
 

--- a/Source/swiftlint/Commands/GenerateDocs.swift
+++ b/Source/swiftlint/Commands/GenerateDocs.swift
@@ -12,7 +12,8 @@ extension SwiftLint {
         var path = "rule_docs"
         @Option(help: "The path to a SwiftLint configuration file")
         var config: String?
-        @OptionGroup var rulesFilterOptions: RulesFilterOptions
+        @OptionGroup
+        var rulesFilterOptions: RulesFilterOptions
 
         func run() throws {
             let configuration = Configuration(configurationFiles: [config].compactMap({ $0 }))

--- a/Source/swiftlint/Commands/Rules.swift
+++ b/Source/swiftlint/Commands/Rules.swift
@@ -17,7 +17,8 @@ extension SwiftLint {
 
         @Option(help: "The path to a SwiftLint configuration file")
         var config: String?
-        @OptionGroup var rulesFilterOptions: RulesFilterOptions
+        @OptionGroup
+        var rulesFilterOptions: RulesFilterOptions
         @Flag(name: .shortAndLong, help: "Display full configuration details")
         var verbose = false
         @Argument(help: "The rule identifier to display description for")

